server/linux_x11/server_x11.py

```diff
@@ -360,114 +389,116 @@ def write_text(text, paste=False, _xdotool=None):
+        elif direction == 'up':
+            libxdo.mouse_up(0, direction)
+        else:
+            raise ValueError('invalid "direction" parameter: "%s"' % direction)
+        time.sleep(delay_millis / 1000)
+
+
+def move_mouse(x, y, reference='absolute', proportional=False, phantom=None):
+    """
+    Move the mouse to the specified coordinates.
+    :param x: x coordinate for move
+    :param y: y coordinate for move
+    :param str reference: One of 'absolute', 'relative' or 'relative_active':
+
+      - absolute: Move the mouse to the absolute position x, y.  x and y will
+        be set to 0 if they are negative.
+      - relative: Move the mouse relative to its current location. x and y may
+        be negative integers.
+      - relative_active: Move the mouse relative to the current active window.
+        0,0 being the top left corner of with window.
+
+    :param proportional:
+    :param phantom: If provided, this parameter should be one of
+      _MOUSE_BUTTONS.keys(). Move to the desired location, click the <phantom>
+      button and restore the mouse to the original location.
+    :type phantom: str or None
+    :return: This function always returns None.
+    """
+    original_location = libxdo.get_mouse_location()
     geo = get_geometry()
+
     if proportional:
         x = geo['width'] * x
         y = geo['height'] * y
-    command = _MOUSE_MOVE_COMMANDS[reference]
-    if command == 'mousemove_active':
-        command = 'mousemove --window %i' % get_active_window()[0]
 
-    if x <= 0 or y <= 0:
-        commands = ['%s -- %f %f' % (command, x, y)]
+    if reference == 'absolute':
+        x = x if x > 0 else x
+        y = y if y > 0 else y
+        libxdo.move_mouse(x, y)
+    elif reference == 'relative_active':
+        window_location = libxdo.get_window_location(libxdo.get_active_window())
+        libxdo.move_mouse(window_location.x + x, window_location.y + y)
+    elif reference == 'relative':
+        libxdo.move_mouse_relative(x, y)
     else:
-        commands = ['%s %f %f' % (command, x, y)]
+        raise ValueError('invalid "reference" parameter "%s"' % reference)
+
     if phantom is not None:
-        commands.append('click %s' % _MOUSE_BUTTONS[phantom])
-        commands.append('mousemove restore')
-
-    # To avoid headaches down the road with argparse, we don't chain commands
-    # if we need to use -- since it would block future flags from being
-    # interpreted.
-    if _xdotool is not None and x >= 0 and y >= 0:
-        _xdotool.extend(commands)
-    else:
-        flush_xdotool(_xdotool)
-        run_command(' '.join(commands))
+        libxdo.click_window(0, _MOUSE_BUTTONS[phantom])
+        libxdo.move_mouse(original_location.x, original_location.y)
 
 
-def pause(amount, _xdotool=None):
-    '''pause amount in ms.'''
-    if _xdotool is not None:
-        _xdotool.append('sleep %f' % (amount / 1000.))
-    else:
-        time.sleep(amount / 1000.)
+def pause(millis):
```

Comment:
millis is a better name than amount, but if we change it here we need to change it everywhere, including in the test scripts. Either way is fine with me. I'm not too worried about grammars calling the RPC methods directly.